instance/models/appserver.py

```diff
@@ -0,0 +1,254 @@
+        is_steady_state = False
+
+    class ConfiguringServer(AppServerState):
+        """ Running Ansible playbooks on server """
+        state_id = 'configuring'
+        name = 'Configuring server'
+        is_steady_state = False
+
+    class Running(AppServerState):
+        """ Instance is up and running """
+        state_id = 'running'
+
+    class ConfigurationFailed(AppServerState):
+        """ Instance was not configured successfully (but may be partially online) """
+        state_id = 'failed'
+        name = 'Configuration failed'
+        is_healthy_state = False
+
+    class Error(AppServerState):
+        """ Instance never got up and running (something went wrong when trying to build new VM) """
+        state_id = 'error'
+        is_healthy_state = False
+
+    class Terminated(AppServerState):
+        """ Instance was running successfully and has been shut down """
+        state_id = 'terminated'
+
+
+# Models ######################################################################
+
+
+class AppServer(ValidateModelMixin, TimeStampedModel):
+    """
+    AppServer - One or more distinct web applications running on a single VM.
+
+    Owned by an Instance.
+
+    Characteristics of an AppServer:
+    * An AppServer instance's configuration fields are *immutable*. If you want to change
+      configuration, change hte Instance's configuration and have it create a new AppServer.
+    * An AppServer owns exactly one VM (Server), onto which it installs its applications.
+    """
+    Status = Status
+    status = ModelResourceStateDescriptor(
+        state_classes=Status.states, default_state=Status.New, model_field_name='_status'
+    )
+    _status = models.CharField(
+        max_length=20,
+        default=status.default_state_class.state_id,
+        choices=status.model_field_choices,
+        db_index=True,
+        db_column='status',
+    )
+    # State transitions:
+    _status_to_waiting_for_server = status.transition(
+        from_states=Status.New, to_state=Status.WaitingForServer
+    )
+    _status_to_configuring_server = status.transition(
+        from_states=Status.WaitingForServer, to_state=Status.ConfiguringServer
+    )
+    _status_to_error = status.transition(
+        from_states=Status.WaitingForServer, to_state=Status.Error
+    )
+    _status_to_running = status.transition(
+        from_states=Status.ConfiguringServer, to_state=Status.Running
+    )
+    _status_to_configuration_failed = status.transition(
+        from_states=Status.ConfiguringServer, to_state=Status.ConfigurationFailed
+    )
+    _status_to_terminated = status.transition(
+        from_states=Status.Running, to_state=Status.Terminated
+    )
+
+    name = models.CharField(max_length=250, blank=False)
+    server = models.OneToOneField(OpenStackServer, on_delete=models.CASCADE, related_name='+')
+    # The Instance that owns this. Instance will get related_name accessors like 'openedxappserver_set'
```

Comment:
@itsjeyd  I see your point. Changed to "InstanceReference has related_name accessors like 'openedxappserver_set'"